Refactor CropModel to infer num_classes from checkpoint during loading

naxatra2 opened the pull request:
Fixes #960

This PR makes `num_classes` optional in the `CropModel` class and infers it from the checkpoint when loading via `load_from_checkpoint`. This improves usability, especially for users who do not know or remember the original number of classes at the time of inference.

**Changes Made**
1.  Made `num_classes` an optional argument in the CropModel constructor.
2.  In the constructor, if `num_classes` is not provided (None), we defer model/metrics initialization until on_load_checkpoint. Otherwise, we initialize them immediately.
3.  Added an `on_load_checkpoint` method that reads `num_classes` from the checkpoint if it is still `None`, then sets up the model and metrics.
4. Added tests to verify that:
   - We can load from a checkpoint without passing `num_classes`.
   - Users can still pass `num_classes` explicitly if they want.
   - Calling `forward` without having a valid model raises a clear error.

bw4sz commented:
This is important and should be merged, two additions. We need to use

```
self.save_hyperparameters()
```

at the end of init documented here https://lightning.ai/docs/pytorch/1.6.2/common/hyperparameters.html#lightningmodule-hyperparameters

The second is the label dict, the crop model uses torchvision dataloader that loads folders from file, that could easily not be the same as the label_dict that as user could optionally pass. We should remove label_dict as an argument. Then we should add a label_dict argument here 

https://github.com/weecology/DeepForest/blob/f1c724dcd48dbfda1fdb5a052c97ee7f59160b5c/src/deepforest/model.py#L157

I believe its under self.train_ds. class_to_idx.

https://pytorch.org/vision/main/generated/torchvision.datasets.DatasetFolder.html#torchvision.datasets.DatasetFolder

Now, the question I have is if a use creates a model, saves the hyperparameters, loads a dataset, saves a checkpoint, and then reloads the model, is the label_dict still there?

naxatra2 commented:
If it is not urgent then can I try doing this @bw4sz, it will be a great learning opportunity. I will update my progress here.

naxatra2 commented:
I have tried to do some changes. mainly did these things:
- [x] changed `self.save_hyperparameters()`
- [x] removed the if statement from here 
> There should not be a if statement here. CropModels should have label dicts and they should return labels, not numeric.
- [x] Test for a checkpoint with explicit `num_classes` to test backward compatibility

----

> Now, the question I have is if a use creates a model, saves the hyperparameters, loads a dataset, saves a checkpoint, and then reloads the model, is the label_dict still there?

Currently, if a user creates a model, loads a dataset (which assigns the label_dict), saves a checkpoint, and then reloads the model, the label_dict is not automatically restored, because it isn’t passed as an init argument

To make the label mapping consistent across checkpoints, we can override the checkpoint hooks to include the label_dict in the saved checkpoint and then restore them when the model is reloaded.

bw4sz commented:
Yes, we can use on_load_checkpoint.

Two other things I saw that generally improve the CropModel

Softmax, not sigmoid, they are equivalent in two class
https://github.com/weecology/DeepForest/blob/e3fd2d087fb8cc69a920e2489f0a5f762436cd96/src/deepforest/model.py#L211

Default value should not be true for validation shuffle.
https://github.com/weecology/DeepForest/blob/e3fd2d087fb8cc69a920e2489f0a5f762436cd96/src/deepforest/model.py#L237

bw4sz commented:
Also, can we make the naming of the crops more informative, 

```
                image_basename = os.path.splitext(os.path.basename(image))[0]
                img_path = os.path.join(savedir, label, f"{image_basename}_{index}.png")
```

https://github.com/weecology/DeepForest/blob/e3fd2d087fb8cc69a920e2489f0a5f762436cd96/src/deepforest/model.py#L200

naxatra2 commented:
@bw4sz are these changes good ? I have made the necessary changes in the last commit

bw4sz commented:
Yes, this is excellent PR, thank you.
